File: results/2026-06-26_103905/tasks/16-environment-matrix-generator/typescript-bun-opus48-1m-high/generated-code/matrix-generator.ts
// matrix-generator.ts
//
// Core library for generating a GitHub Actions `strategy.matrix` build matrix
// from a declarative configuration. Built incrementally with red/green TDD.
//
// Step #1 (GREEN): expand the named axes into the full cartesian product.

/** A scalar value allowed inside a matrix combination. */
export type MatrixValue = string | number | boolean;

/** A single fully-resolved combination, e.g. { os: "ubuntu-latest", language: "20" }. */
export type Combination = Record<string, MatrixValue>;

/** The matrix sub-block of a GitHub Actions strategy. */
export interface StrategyMatrix {
  [axis: string]: MatrixValue[] | Combination[] | undefined;
  include?: Combination[];
  exclude?: Combination[];
}

/** A GitHub Actions `jobs.<id>.strategy` object. */
export interface Strategy {
  "fail-fast": boolean;
  "max-parallel"?: number;
  matrix: StrategyMatrix;
}

/** Result of expanding a matrix configuration. */
export interface MatrixResult {
  /** GitHub-Actions-ready strategy block (drop straight into a workflow). */
  strategy: Strategy;
  /** Fully expanded list of build combinations (after exclude/include rules). */
  combinations: Combination[];
  /** Number of combinations (== combinations.length). */
  count: number;
}

/**
 * Thrown when the expanded matrix has more combinations than `maxSize` allows.
 * Carries the offending sizes so callers can render a meaningful message.
 */
export class MatrixSizeError extends Error {
  constructor(
    public readonly actual: number,
    public readonly maxSize: number,
  ) {
    super(
      `Matrix has ${actual} combinations, which exceeds the maximum of ${maxSize}. ` +
        `Reduce the number of axis values, add exclude rules, or raise maxSize.`,
    );
    this.name = "MatrixSizeError";
  }
}

/** Declarative matrix configuration (the input to {@link generateMatrix}). */
export interface MatrixConfig {
  /** Operating system axis. */
  os?: MatrixValue[];
  /** Language / runtime version axis. */
  language?: MatrixValue[];
  /** Feature-flag axis. */
  features?: MatrixValue[];
  /** Combinations to remove (partial match against expanded combinations). */
  exclude?: Combination[];
  /** Combinations to add/extend (GitHub Actions `matrix.include` semantics). */
  include?: Combination[];
  /** Limit on concurrently-running jobs (GitHub `strategy.max-parallel`). */
  maxParallel?: number;
  /** Whether to cancel in-progress jobs if any matrix job fails (default true). */
  failFast?: boolean;
  /** Hard cap on the number of expanded combinations (default 256, GitHub's limit). */
  maxSize?: number;
}

/** GitHub Actions itself caps a matrix at 256 combinations. */
export const DEFAULT_MAX_SIZE = 256;

/** Thrown when the configuration is structurally invalid. */
export class MatrixConfigError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "MatrixConfigError";
  }
}

/**
 * Return true if `combo` matches `pattern`: every key in `pattern` is present
 * in `combo` with an equal value. Extra keys in `combo` are ignored, so a
 * partial pattern (e.g. just `{ os }`) matches many combinations.
 */
function matchesPattern(combo: Combination, pattern: Combination): boolean {
  return Object.keys(pattern).every((key) => combo[key] === pattern[key]);
}

/** Remove every combination matching any exclude entry. */
function applyExclude(combinations: Combination[], excludes: Combination[]): Combination[] {
  if (excludes.length === 0) return combinations;
  return combinations.filter(
    (combo) => !excludes.some((pattern) => matchesPattern(combo, pattern)),
  );
}

/**
 * Compute the cartesian product of a set of named axes, preserving the order
 * in which axes are declared and the order of values within each axis. This
 * matches GitHub Actions' expansion order (first axis varies slowest).
 */
function cartesianProduct(axes: Record<string, MatrixValue[]>): Combination[] {
  const keys = Object.keys(axes);
  // With no axes there are no base combinations at all (an include-only matrix
  // relies on this so each include entry becomes a standalone combination,
  // rather than everything merging into one empty seed combination).
  if (keys.length === 0) return [];
  // Seed with a single empty combination, then fan out over each axis.
  let combinations: Combination[] = [{}];
  for (const key of keys) {
    const values = axes[key];
    const next: Combination[] = [];
    for (const combo of combinations) {
      for (const value of values) {
        next.push({ ...combo, [key]: value });
      }
    }
    combinations = next;
  }
  return combinations;
}

/**
 * Collect the declared named axes (os/language/features) into an ordered
 * axis map, skipping any axis that is absent or empty.
 */
function collectAxes(config: MatrixConfig): Record<string, MatrixValue[]> {
  const axes: Record<string, MatrixValue[]> = {};
  for (const key of ["os", "language", "features"] as const) {
    const values = config[key];
    if (values && values.length > 0) {
      axes[key] = values;
    }
  }
  return axes;
}

/**
 * Apply GitHub Actions `include` semantics.
 *
 * For each include entry, in order:
 *  - It can extend an expanded combination only if it does NOT overwrite any
 *    of that combination's ORIGINAL axis values (non-axis keys may differ).
 *  - It is added to every combination it can extend.
 *  - If it cannot extend ANY combination, it is appended as a brand-new
 *    standalone combination.
 *
 * Only the original cartesian-product combinations are eligible for
 * extension; combinations created by earlier include entries are not.
 */
function applyInclude(
  combinations: Combination[],
  includes: Combination[],
  axisKeys: string[],
): Combination[] {
  if (includes.length === 0) return combinations;

  const axisKeySet = new Set(axisKeys);
  // Mark which entries come from the original expansion (extendable).
  const result: Combination[] = combinations.map((c) => ({ ...c }));
  const originalCount = result.length;

  for (const include of includes) {
    let matched = false;
    for (let i = 0; i < originalCount; i++) {
      const combo = result[i];
      // The include overwrites an original axis value if it sets an axis key
      // to a value different from the combination's current value.
      const overwritesAxis = Object.keys(include).some(
        (key) => axisKeySet.has(key) && combo[key] !== include[key],
      );
      if (overwritesAxis) continue;
      // Eligible: merge the include's keys in (added values may be overwritten).
      result[i] = { ...combo, ...include };
      matched = true;
    }
    if (!matched) {
      // Could not extend any combination -> standalone new combination.
      result.push({ ...include });
    }
  }
  return result;
}

/** True for plain objects (not arrays, not null). */
function isPlainObject(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

/** Validate that a value is an array of plain objects (for include/exclude). */
function assertCombinationArray(value: unknown, field: string): asserts value is Combination[] {
  if (!Array.isArray(value)) {
    throw new MatrixConfigError(`"${field}" must be an array of objects.`);
  }
  value.forEach((entry, i) => {
    if (!isPlainObject(entry)) {
      throw new MatrixConfigError(`"${field}[${i}]" must be an object.`);
    }
  });
}

/**
 * Validate and narrow raw, untrusted input (e.g. parsed JSON) into a
 * {@link MatrixConfig}. Throws {@link MatrixConfigError} with a clear message
 * for any structural problem so the CLI can report it gracefully.
 */
export function parseConfig(raw: unknown): MatrixConfig {
  if (!isPlainObject(raw)) {
    throw new MatrixConfigError("Configuration must be a JSON object.");
  }

  const config: MatrixConfig = {};

  // Validate the three named axes.
  for (const key of ["os", "language", "features"] as const) {
    if (raw[key] === undefined) continue;
    if (!Array.isArray(raw[key])) {
      throw new MatrixConfigError(`Axis "${key}" must be an array of values.`);
    }
    config[key] = raw[key] as MatrixValue[];
  }

  if (raw.include !== undefined) {
    assertCombinationArray(raw.include, "include");
    config.include = raw.include;
  }
  if (raw.exclude !== undefined) {
    assertCombinationArray(raw.exclude, "exclude");
    config.exclude = raw.exclude;
  }

  if (raw.maxParallel !== undefined) {
    if (typeof raw.maxParallel !== "number") {
      throw new MatrixConfigError(`"maxParallel" must be a number.`);
    }
    config.maxParallel = raw.maxParallel;
  }
  if (raw.failFast !== undefined) {
    if (typeof raw.failFast !== "boolean") {
      throw new MatrixConfigError(`"failFast" must be a boolean.`);
    }
    config.failFast = raw.failFast;
  }
  if (raw.maxSize !== undefined) {
    if (typeof raw.maxSize !== "number") {
      throw new MatrixConfigError(`"maxSize" must be a number.`);
    }
    config.maxSize = raw.maxSize;
  }

  return config;
}

/** Validate semantic constraints on an already-typed config. */
function validateConfig(config: MatrixConfig): void {
  const hasAxes = (["os", "language", "features"] as const).some(
    (k) => config[k] && config[k]!.length > 0,
  );
  const hasInclude = (config.include?.length ?? 0) > 0;
  if (!hasAxes && !hasInclude) {
    throw new MatrixConfigError(
      "Configuration must define at least one axis (os/language/features) or an include list.",
    );
  }
  if (config.maxParallel !== undefined && config.maxParallel < 1) {
    throw new MatrixConfigError(`"maxParallel" must be a positive integer, got ${config.maxParallel}.`);
  }
  if (config.maxSize !== undefined && config.maxSize < 1) {
    throw new MatrixConfigError(`"maxSize" must be a positive integer, got ${config.maxSize}.`);
  }
}

/** Generate the expanded build matrix from a configuration. */
export function generateMatrix(config: MatrixConfig): MatrixResult {
  validateConfig(config);
  const axes = collectAxes(config);
  const axisKeys = Object.keys(axes);
  let combinations = cartesianProduct(axes);
  combinations = applyExclude(combinations, config.exclude ?? []);
  combinations = applyInclude(combinations, config.include ?? [], axisKeys);

  // Validate size BEFORE returning so callers never see an over-large matrix.
  const maxSize = config.maxSize ?? DEFAULT_MAX_SIZE;
  if (combinations.length > maxSize) {
    throw new MatrixSizeError(combinations.length, maxSize);
  }

  // Assemble the GitHub-Actions-ready strategy block.
  const matrix: StrategyMatrix = { ...axes };
  if (config.include && config.include.length > 0) matrix.include = config.include;
  if (config.exclude && config.exclude.length > 0) matrix.exclude = config.exclude;

  const strategy: Strategy = {
    "fail-fast": config.failFast ?? true,
    matrix,
  };
  if (config.maxParallel !== undefined) {
    strategy["max-parallel"] = config.maxParallel;
  }

  return { strategy, combinations, count: combinations.length };
}

File: results/2026-07-01_184135/tasks/16-environment-matrix-generator/default-fable5-high/generated-code/matrix_generator.py
"""Generate a GitHub Actions strategy.matrix from a declarative config.

Approach
--------
The generator mirrors GitHub Actions' own matrix semantics:

1. Take the cartesian product of every axis (os, language versions, flags...).
2. Apply ``exclude`` rules: a combination is removed when an exclude entry is
   a *partial* match (every key/value in the entry matches the combination).
3. Apply ``include`` rules: an include entry is merged into every product
   combination it matches without overwriting original axis values; entries
   that match nothing become brand-new standalone combinations.
4. Validate the final matrix against a maximum size.

The result is emitted as the standard "dynamic matrix" JSON shape consumed by
``strategy.matrix: ${{ fromJSON(...) }}`` — i.e. ``{"include": [...]}`` plus
``fail-fast`` / ``max-parallel`` strategy settings.
"""
import json
import sys
from itertools import product

# GitHub Actions caps a matrix at 256 jobs; we use the same default.
DEFAULT_MAX_SIZE = 256

# Types allowed as axis values (what YAML scalars map to in JSON).
SCALAR_TYPES = (str, int, float, bool)


class MatrixConfigError(ValueError):
    """Raised when the configuration is invalid or the matrix is too large."""


def _validate_config(config):
    """Check the config document's shape, raising MatrixConfigError early
    with a message that names the offending field."""
    if not isinstance(config, dict):
        raise MatrixConfigError("config must be a JSON object")

    matrix = config.get("matrix")
    if not isinstance(matrix, dict) or not matrix:
        raise MatrixConfigError(
            "config must contain a non-empty 'matrix' object of axes"
        )

    for key, value in matrix.items():
        if key in ("include", "exclude"):
            if not (
                isinstance(value, list)
                and all(isinstance(entry, dict) for entry in value)
            ):
                raise MatrixConfigError(
                    f"'{key}' must be a list of objects"
                )
            continue
        if not isinstance(value, list) or not value:
            raise MatrixConfigError(
                f"matrix axis '{key}' must be a non-empty list of values"
            )
        for item in value:
            if not isinstance(item, SCALAR_TYPES):
                raise MatrixConfigError(
                    f"matrix axis '{key}' contains a non-scalar value: {item!r}"
                )

    if not any(key not in ("include", "exclude") for key in matrix):
        raise MatrixConfigError(
            "config must contain a non-empty 'matrix' object of axes"
        )

    if not isinstance(config.get("fail-fast", True), bool):
        raise MatrixConfigError("'fail-fast' must be a boolean")

    for field, default in (("max-parallel", 1), ("max-size", DEFAULT_MAX_SIZE)):
        value = config.get(field, default)
        if not isinstance(value, int) or isinstance(value, bool) or value < 1:
            raise MatrixConfigError(f"'{field}' must be a positive integer")


def expand_matrix(matrix):
    """Return the cartesian product of the matrix axes as a list of dicts.

    Axis order (and value order within an axis) is preserved, matching the
    order GitHub Actions enumerates jobs in.
    """
    keys = list(matrix)
    return [dict(zip(keys, values)) for values in product(*matrix.values())]


def apply_excludes(combos, excludes):
    """Drop every combination partially matched by an exclude entry.

    GitHub Actions semantics: an exclude entry removes a combination when
    *all* of the entry's key/value pairs match it — the entry does not need
    to specify every axis.
    """

    def is_excluded(combo):
        return any(
            all(combo.get(key) == value for key, value in entry.items())
            for entry in excludes
        )

    return [combo for combo in combos if not is_excluded(combo)]


def apply_includes(combos, includes, original_keys):
    """Merge include entries into the expanded matrix (GitHub semantics).

    For each include entry:
      * It "matches" a combination when none of its pairs would overwrite an
        original axis value (pairs for original keys must be equal; pairs for
        keys previous includes added may be overwritten freely).
      * Matching entries have their extra pairs merged into every match.
      * An entry matching no combination is appended as a new combination.
    """
    original_keys = set(original_keys)
    result = [dict(combo) for combo in combos]

    for entry in includes:
        matched = False
        for combo in result:
            # An original combination is only expandable, never mutated on
            # its axis values: every entry pair naming an original key must
            # equal the combo's value for the entry to apply.
            if all(
                combo.get(key) == value
                for key, value in entry.items()
                if key in original_keys
            ):
                matched = True
                for key, value in entry.items():
                    if key not in original_keys:
                        combo[key] = value
        if not matched:
            result.append(dict(entry))
    return result


def generate(config):
    """Build the full strategy document from a config dict.

    Returns ``{"fail-fast": ..., "max-parallel": ..., "count": N,
    "matrix": {"include": [combos]}}`` — the shape a workflow consumes via
    ``strategy: ${{ fromJSON(needs.gen.outputs.strategy) }}`` or by reading
    the ``matrix`` member alone.
    """
    _validate_config(config)

    matrix = dict(config["matrix"])
    includes = matrix.pop("include", [])
    excludes = matrix.pop("exclude", [])

    combos = expand_matrix(matrix)
    combos = apply_excludes(combos, excludes)
    if not combos:
        raise MatrixConfigError(
            "matrix is empty after applying exclude rules; relax the "
            "'exclude' entries or add axis values"
        )
    combos = apply_includes(combos, includes, matrix.keys())

    max_size = config.get("max-size", DEFAULT_MAX_SIZE)
    if len(combos) > max_size:
        raise MatrixConfigError(
            f"matrix of {len(combos)} combinations exceeds the maximum of "
            f"{max_size}; shrink the axes or raise 'max-size'"
        )

    result = {"fail-fast": config.get("fail-fast", True)}
    if "max-parallel" in config:
        result["max-parallel"] = config["max-parallel"]
    result["count"] = len(combos)
    result["matrix"] = {"include": combos}
    return result


def main(argv=None):
    """CLI entry point: read a config file, print the matrix as one line of
    compact JSON (easy to pipe into ``$GITHUB_OUTPUT``). Returns the exit
    code instead of raising, so tests can call it in-process."""
    argv = sys.argv[1:] if argv is None else argv
    if len(argv) != 1:
        print(
            "usage: python3 matrix_generator.py <config.json>",
            file=sys.stderr,
        )
        return 2

    path = argv[0]
    try:
        with open(path, encoding="utf-8") as handle:
            raw = handle.read()
    except OSError as exc:
        print(f"error: cannot read config file '{path}': {exc}", file=sys.stderr)
        return 1

    try:
        config = json.loads(raw)
    except json.JSONDecodeError as exc:
        print(f"error: '{path}' is not valid JSON: {exc}", file=sys.stderr)
        return 1

    try:
        strategy = generate(config)
    except MatrixConfigError as exc:
        print(f"error: invalid matrix config: {exc}", file=sys.stderr)
        return 1

    # Compact separators keep the document on one shell-safe line.
    print(json.dumps(strategy, separators=(",", ":")))
    return 0


if __name__ == "__main__":
    sys.exit(main())
